src/timer.rs

```diff
@@ -0,0 +1,285 @@
+        }
+
+        #[must_use]
+        pub fn divider(mut self, divider: u32) -> Self {
+            self.divider = divider;
+            self
+        }
+
+        #[must_use]
+        #[cfg(any(esp32s2, esp32s3, esp32c3))]
+        pub fn xtal(mut self, xtal: bool) -> Self {
+            self.xtal = xtal;
+            self
+        }
+    }
+
+    impl Default for Config {
+        fn default() -> Self {
+            Self {
+                divider: 2,
+                #[cfg(any(esp32s2, esp32s3, esp32c3))]
+                xtal: false,
+            }
+        }
+    }
+}
+
+pub trait Timer: Send {
+    fn group() -> timer_group_t;
+    fn index() -> timer_idx_t;
+}
+
+pub struct TimerDriver<'d, TIMER>
+where
+    TIMER: Timer,
+{
+    _timer: PeripheralRef<'d, TIMER>,
+}
+
+impl<'d, TIMER> TimerDriver<'d, TIMER>
+where
+    TIMER: Timer,
+{
+    pub fn new(
+        timer: impl Peripheral<P = TIMER> + 'd,
+        config: &config::Config,
+    ) -> Result<TimerDriver<'d, TIMER>, EspError> {
+        crate::into_ref!(timer);
+
+        esp!(unsafe {
+            timer_init(
+                TIMER::group(),
+                TIMER::index(),
+                &timer_config_t {
+                    alarm_en: timer_alarm_t_TIMER_ALARM_DIS,
+                    counter_en: timer_start_t_TIMER_PAUSE,
+                    counter_dir: timer_count_dir_t_TIMER_COUNT_UP,
+                    auto_reload: timer_autoreload_t_TIMER_AUTORELOAD_DIS,
+                    intr_type: timer_intr_mode_t_TIMER_INTR_LEVEL,
+                    divider: config.divider,
+                    #[cfg(any(esp32s2, esp32s3, esp32c3))]
+                    clk_src: if config.xtal {
+                        timer_src_clk_t_TIMER_SRC_CLK_XTAL
+                    } else {
+                        timer_src_clk_t_TIMER_SRC_CLK_ARB
+                    },
+                },
+            )
+        })?;
+
+        Ok(TimerDriver { _timer: timer })
+    }
+
+    pub fn enable(&mut self, enable: bool) -> Result<(), EspError> {
+        if enable {
+            esp!(unsafe { timer_start(TIMER::group(), TIMER::index()) })?;
+        } else {
+            esp!(unsafe { timer_pause(TIMER::group(), TIMER::index()) })?;
+        }
+
+        Ok(())
+    }
+
+    pub fn counter(&self) -> Result<u64, EspError> {
+        let mut value = 0_u64;
+
+        esp!(unsafe {
+            timer_get_counter_value(TIMER::group(), TIMER::index(), &mut value as *mut _)
+        })?;
+
+        Ok(value)
+    }
+
+    pub fn set_counter(&mut self, value: u64) -> Result<(), EspError> {
+        esp!(unsafe { timer_set_counter_value(TIMER::group(), TIMER::index(), value) })?;
+
+        Ok(())
+    }
+
+    pub fn enable_alarm(&mut self, enable: bool) -> Result<(), EspError> {
+        esp!(unsafe {
+            timer_set_alarm(
+                TIMER::group(),
+                TIMER::index(),
+                if enable {
+                    timer_alarm_t_TIMER_ALARM_DIS
```

Comment:
Should this perhaps be
```suggestion
                    timer_alarm_t_TIMER_ALARM_EN
```